adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/providers/ActilityProvider.java

```diff
@@ -87,145 +86,98 @@ protected Buffer extractPayload(final JsonObject loraMessage) {
-                        LoraUtils.getChildObject(lrr, FIELD_ACTILITY_LRR_RSSI, String.class)
-                            .ifPresent(s -> normalizedGateway.put(LoraConstants.APP_PROPERTY_RSS, Math.abs(Double.valueOf(s))));
-                        LoraUtils.getChildObject(lrr, FIELD_ACTILITY_LRR_SNR, String.class)
-                        .ifPresent(s -> normalizedGateway.put(LoraConstants.APP_PROPERTY_SNR, Math.abs(Double.valueOf(s))));
-                        return normalizedGateway;
-                    })
-                    .collect(() -> new JsonArray(), (array, value) -> array.add(value), (array1, array2) -> array1.addAll(array2));
-                data.put(LoraConstants.GATEWAYS, normalizedGateways.toString());
+                    .map(this::extractGatewayInfo)
+                    .forEach(gateway -> {
+                        Optional.ofNullable(gateway.getGatewayId())
+                            .ifPresent(s -> gwId.ifPresent(id -> {
+                                if (id.equals(s)) {
+                                    Optional.ofNullable(LoraUtils.newLocationFromString(
+                                            LoraUtils.getChildObject(rootObject, FIELD_ACTILITY_LONGITUDE, String.class),
+                                            LoraUtils.getChildObject(rootObject, FIELD_ACTILITY_LATITUTDE, String.class),
+                                            Optional.empty()))
+                                        .ifPresent(gateway::setLocation);
+                                }
+                            }));
+                        data.addGatewayInfo(gateway);
+                    });
             });
-
         return data;
     }
 
-    @Override
-    protected JsonObject extractAdditionalData(final JsonObject loraMessage) {
-        final JsonObject returnMessage = loraMessage.copy();
-        if (returnMessage.containsKey(FIELD_ACTILITY_LRR_RSSI)) {
-            returnMessage.remove(LoraConstants.APP_PROPERTY_RSS);
-        }
-        if (returnMessage.containsKey(FIELD_ACTILITY_TX_POWER)) {
-            returnMessage.remove(LoraConstants.APP_PROPERTY_TX_POWER);
-        }
-        if (returnMessage.containsKey(FIELD_ACTILITY_CHANNEL)) {
-            returnMessage.remove(LoraConstants.APP_PROPERTY_CHANNEL);
-        }
-        if (returnMessage.containsKey(FIELD_ACTILITY_SUB_BAND)) {
-            returnMessage.remove(LoraConstants.APP_PROPERTY_SUB_BAND);
+    /**
+     * Gets the frequency corresponding to the channel ID used by Actility/ThingWork
+     * as described in section 2.4 of the
+     * <a href="https://partners.thingpark.com/sites/default/files/2017-11/AdvancedThingParkDeveloperGuide_V4.pdf">
+     * Advanced Developer Guide</a>.
+     *
+     * @param logicalChannelId The channel ID.
+     * @return The frequency in MHz or {@code null} if the identifier is unknown.
+     */
+    private Double getFrequency(final String logicalChannelId) {
+        switch (logicalChannelId) {
```

Comment:
I would say, things like this belong into an enum.

Reply:
In general, I agree. However, I haven't found a (formal) definition of these channel IDs and I suspect that they are specific to each provider (if used at all). We could define a private enum but is that worth the effort?

Reply:
Ok, in that case leaving it as is makes sense.